common/src/main/java/io/netty/util/concurrent/FastThreadLocal.java

```diff
@@ -147,19 +150,26 @@ public final V get() {
 
     private void registerCleaner(final InternalThreadLocalMap threadLocalMap) {
         Thread current = Thread.currentThread();
-        if (!FastThreadLocalThread.willCleanupFastThreadLocals(current)) {
-            // We will need to ensure we will trigger remove(InternalThreadLocalMap) so everything will be released
-            // and FastThreadLocal.onRemoval(...) will be called.
-            ObjectCleaner.register(current, new Runnable() {
-                @Override
-                public void run() {
-                    remove(threadLocalMap);
-
-                    // It's fine to not call InternalThreadLocalMap.remove() here as this will only be triggered once
-                    // the Thread is collected by GC. In this case the ThreadLocal will be gone away already.
-                }
-            });
+        if (FastThreadLocalThread.willCleanupFastThreadLocals(current)) {
+            return;
         }
+        Object v = threadLocalMap.indexedVariable(cleanerFlagIndex);
+        if (v != InternalThreadLocalMap.UNSET) {
+            return;
+        }
+        threadLocalMap.setIndexedVariable(cleanerFlagIndex, Boolean.TRUE);
```

Comment:
consider adding a comment that `threadLocalMap.removeIndexedVariable(cleanerFlagIndex);` isn't necessary because the finally cleanup is tied to the lifetime of the thread, and this Object will be discarded if the associated thread is GCed.